### PR TITLE
fix and improve detection and import of ods, xlsx and csv documents

### PR DESCRIPTION
**Pull request by blizzz**
addresses #1440

This PR fixes a few corner cases with importing from different document types, especially about recognization of date formats (but not only). For instance, it is not possible again to re-import and exported CSV.

Integration tests are extended with column formats and do not only test csv import anymore, but also ODS, XLSX (from 365) and XLSX (from LibreOffice) – they have some subtle differences and failed differently along the way.

---
#### Historic notes

<details><summary>With the first commit, the added test fails as expected.</summary>

Screenshot_20241111_175746
</details>

<details><summary>The second commit brings in two tests for a similar table in xlsx files (by SharePoint/365 and by LibreOffice). The first tests succeeds with a little datetime adjustment – containing the time, although only the date was entered as value. Maybe it is possible to read out the format and adjust automatically. The second test against the LibreOffice-created file fails with import errors. This needs further investigation.

**P.S.:** Changed back to date-only type.
</summary>

Screenshot_20241113_181325

</details>

The third commit adds the treatment of date-only datetime types imported from native xlsx.

<details><summary>The eights commit also fixes the rendering of the value in the preview dialog. For instance in this screenshot, the Date used to be shown as Excel timestamp.</summary>

Screenshot_20241115_173756

</details>

### CODE REVIEW

**Comment by blizzz** [comment]
The PHP 8.4 tests needs a phpunit compat update. Maybe it'll be there after this weekend via dependabot and we can rebase.

File: lib/Service/ImportService.php [hunk 1]
```diff
@@ -439,6 +482,12 @@ private function getColumns(Row $firstRow, Row $secondRow): void {
 				$dataTypes[] = $this->parseColumnDataType($secondRowCellIterator->current());
 			} else {
 				$this->logger->debug('No cell given or cellValue is empty while loading columns for importing');
+				if ($cell->getDataType() === 'null') {
+					// LibreOffice generated XLSX doc may have more empty columns in the first row.
+					// Continue without increasing error count.
+					// Question: What about tables where a column does not have a heading?
```
[the remark this answers is not in the document]
Comment (blizzz) [reply lib/Service/ImportService.php:488]: Ha. Good finding. Forgot about this one. Let me check me one thing, to ensure that this does not change from the previous behavior. Might indeed need a touch. 

There is also a documentation component in that headers were expected on all columns.

[the remark this answers is not in the document]
Comment (blizzz) [reply lib/Service/ImportService.php:488]: Okay, import works as before. But there is one difference:

Screenshot_20241126_123404

Previously it was one. 


It was testing with a three-column table where the middle-one was not having a caption. But the values were set (always strings for simplicity). Before the change and now, only the columns with a caption were being added, the middle one ignored.

On import, however, the first two columns were added and the third column ignored. 

To visualize, the original

| *Caption 1* | | *Caption 3* |
| --- | --- | --- |
| *Aaa* | *A* | Zzz |
| *Bbb* | *B* | Yyy |
| *Ccc* | *C* | Xxx |

turns  (elements taken over in *italic*) into 


| *Caption 1* | *Caption 3* |
| --- | --- |
| *Aaa* | *A* |
| *Bbb* | *B* |
| *Ccc* | *C* |

Our columns require a caption. Something to document, but I look into increasing the error again, if it is not a trailing column.

[the remark this answers is not in the document]
Comment (blizzz) [reply lib/Service/ImportService.php]: Solved in https://github.com/nextcloud/tables/pull/1446/commits/50815ae67bc3a11745de29e70a06c80623b15b6c

[the remark this answers is not in the document]
Comment (blizzz) [reply lib/Service/ImportService.php]: … and added https://github.com/nextcloud/tables/wiki/Usage#import

[the remark this answers is not in the document]
Comment (juliusknorr) [reply lib/Service/ImportService.php]: We could also add that as a note to the import dialog, most users will probably never get to the wiki

**Review by juliusknorr — APPROVED** [review]

File: lib/Service/ImportService.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (blizzz) [reply lib/Service/ImportService.php]: > We could also add that as a note to the import dialog, most users will probably never get to the wiki

:+1:  Added a brief supplement in 36179dc13eafa89626e287e9cede138f79b9e835

[the remark this answers is not in the document]
Comment (juliusknorr) [reply lib/Service/ImportService.php]: Thanks 👍
